dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraph.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.dependencies;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.eclipse.aether.artifact.Artifact;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.TreeMultimap;
+
+/**
+ * A representation of the complete non-cyclic transitive dependency tree of a Maven artifact.
+ * 
+ * <p>Imagine performing a breadth first search through the tree. As we go we build up a list of
+ * dependencies. The path to each dependency node is placed in a list. Although each path should
+ * appear only once, each dependency may appear many times in different paths. This representation
+ * is unusual because it represents a tree as a list of every path from the root to each node,
+ * instead of a network of nodes.
+ * 
+ * <p>Artifacts are considered to be the same if they have the same group ID, artifact ID, and version.
+ */
+public class DependencyGraph {
+
+  // DependencyGraphBuilder builds this in breadth first order.
+  // That is, this list contains the paths to each node in breadth first order 
+  private final List<DependencyPath> graph = new ArrayList<>();
+  
+  // map of groupId:artifactId to versions
+  // todo if versions' values were the whole coordinate string 
+  // (or even the Artifact itself), would this be simpler?
+  private final TreeMultimap<String, String> versions =
+      TreeMultimap.create(Comparator.naturalOrder(), new VersionComparator());
+  
+  // map of groupId:artifactId:version to paths
+  private SetMultimap<String, DependencyPath> paths = HashMultimap.create();
+  
+  // hide constructor
+  DependencyGraph() {
```

Comment:
Did you intend to leave accessible to other code in the same package?

Reply:
yes. Here I wanted to force use of the builder.